### PR TITLE
MC-34351: [On-Premise] Configurable product out of stock cart issue

### PR DESCRIPTION
**Pull request by odubovyk**
### Bug
* [MC-34351](https://jira.corp.magento.com/browse/MC-34351) [On-Premise] Configurable product out of stock cart issue

### Description (*)

Configurable product should show on storefront as "out of stock" if its stock status is "Out of Stock" regardless of its options stock statuse

### Related Pull Requests
none

#### Scenario 1 (Default Stock)

1. Set _Catalog > Inventory > Stock Options >>Display Out of Stock Products=Yes_
2. Create configurable product with options with qty > 0 and save changes
3. Check that configurable product is "in stock" on PLP and PDP and can be added to shopping cart
4. Set configurable product stock status to "out of stock" and save changes
5. Visit PLP and PDP on storefront

🔴 Actual Result
Configurable product is "in stock" on PDP
🟢 Expected Result
Configurable product is "out of stock" on PDP and PLP and cannot be added to shopping cart

#### Scenario 2 (Custom Stock)

1. Set _Catalog > Inventory > Stock Options >>Display Out of Stock Products=Yes_
2. Create second inventory source
3. Create second stock and assign main website and second source
4. Create configurable product with options and assign options to second source with qty > 0 and "in stock"
5. Check that configurable product is "in stock" on PLP and PDP and can be added to shopping cart
6. Set configurable product stock status to "out of stock" and save changes
7. Visit PLP and PDP on storefront

🔴 Actual Result
Configurable product is "in stock" on PDP and PLP
🟢 Expected Result
Configurable product is "out of stock" on PDP and PLP and cannot be added to shopping cart

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run functional tests

**Comment by odubovyk** [comment]
@magento run failed tests

**Comment by odubovyk** [comment]
@magento run all tests

**Review by sidolov — CHANGES_REQUESTED** [review]
@odubovyk please, check the same case with additional stock and configurable assigned to it instead of default stock.

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run Integration Tests

**Comment by odubovyk** [comment]
@magento run Functional Tests CE

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by odubovyk** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run Integration Tests

**Comment by bubasuma** [comment]
@magento run Integration Tests

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run all tests

File: InventoryConfigurableProduct/Plugin/CatalogInventory/Helper/Stock/AdaptAssignStatusToProductPlugin.php [hunk 1]
```diff
@@ -76,6 +86,18 @@ public function beforeAssignStatusToProduct(
         if ($product->getTypeId() === Configurable::TYPE_CODE) {
             $website = $this->storeManager->getWebsite();
             $stock = $this->stockResolver->execute(SalesChannelInterface::TYPE_WEBSITE, $website->getCode());
+            $stockId = $stock->getStockId();
+            try {
+                $stockItemData = $this->getStockItemData->execute($product->getSku(), $stockId);
+            } catch (NoSuchEntityException $exception) {
+                $stockItemData = null;
+            }
+            if (null !== $stockItemData) {
```
Comment (sidolov) [inline InventoryConfigurableProduct/Plugin/CatalogInventory/Helper/Stock/AdaptAssignStatusToProductPlugin.php:95]: Remove if statement and move the code to try block

File: InventoryConfigurableProduct/Plugin/CatalogInventory/Helper/Stock/AdaptAssignStatusToProductPlugin.php [hunk 2]
```diff
@@ -76,6 +86,18 @@ public function beforeAssignStatusToProduct(
...
+                $salableConfigurable = (bool)$stockItemData[GetStockItemDataInterface::IS_SALABLE];
```
Comment (sidolov) [inline InventoryConfigurableProduct/Plugin/CatalogInventory/Helper/Stock/AdaptAssignStatusToProductPlugin.php]: Unnecessary variable, please, use `$stockItemData[GetStockItemDataInterface::IS_SALABLE]` instead

**Review by sidolov — CHANGES_REQUESTED** [review]

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run all tests

**Comment by bubasuma** [comment]
@magento run Functional Tests EE

**Comment by bubasuma** [comment]
@magento run Functional Tests EE

**Comment by bubasuma** [comment]
@magento run Functional Tests EE

**Review by sidolov — APPROVED** [review]

**Comment by bubasuma** [comment]
@magento run Functional Tests EE

**Comment by bubasuma** [comment]
@magento run Functional Tests EE

**Comment by engcom-Oscar** [comment]
Hello @odubovyk ,
I have tested this PR and found out that with enabled Display Out of Stock Products - on category page I can see Configurable products, not assigned to present stock. like I have configuration that Main website assigned to custom stock and source, but i see configurable products with inventory on default source (stock) with status 'Out of stock'. May you please recheck it.

**Comment by bubasuma** [comment]
@engcom-Oscar 
I believe you're observing expected behavior. See https://github.com/magento/inventory/pull/3085

**Comment by engcom-Oscar** [comment]
@bubasuma , I don't agree because these products assigned to particular sources and stocks

**Comment by bubasuma** [comment]
@engcom-Oscar The fact is the behavior you're observing now has nothing to do with this fix. You will be able to reproduce the same behavior with simple products in mainline (1.2-develop). I pointed out to the pull request #3085 where you can find more details about this behavior.

cc: @sidolov

**Comment by sidolov** [comment]
@magento import pr to magento-commerce/inventory

**Comment by magento-engcom-team** [comment]
@sidolov the pull request successfully imported.

**Comment by bubasuma** [comment]
@GordonLesti ,
The `IF(inventory_stock_item.is_in_stock = 0, 0, MAX(stock.is_salable))` means if configurable product stock status is "Out Of Stock", then "is_salable" is 0 regardless of the options stock statuses.

`cataloginventory_stock_item. is_in_stock` can be set manually in product edit page for configurable product and it should be prioritized over the options stock statuses in both default stock and custom stock.

Please look at the test cases in the description.

**Comment by GordonLesti** [comment]
Hello @bubasuma ,
thank you very much for the fast reply. But unfortunately the value of `cataloginventory_stock_item.is_in_stock` seems to be automatically updated to `0` on configurable products if their children are out of stock in the default stock. Can that be?
One day we tried to update all products with
```
UPDATE cataloginventory_stock_item SET is_in_stock = 1; 
```
but maybe a day later we had again many entries with value `0`.

Best Regards
Gordon

**Pull request merged**